source/dofs/dof_tools.cc

```diff
@@ -765,6 +765,110 @@ namespace DoFTools
+    };
+
+    std::vector<types::global_dof_index> local_dof_indices;
+    local_dof_indices.reserve (max_dofs_per_cell(dof_handler));
+
+    // Get all the dofs that live on the subdomain:
+    std::set<types::global_dof_index> predicate_dofs;
+
+    typename DoFHandlerType::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+    for (; cell!=endc; ++cell)
+      if (!cell->is_artificial() && predicate(cell))
+        {
+          local_dof_indices.resize (cell->get_fe().dofs_per_cell);
+          cell->get_dof_indices (local_dof_indices);
+          predicate_dofs.insert(local_dof_indices.begin(),
+                                local_dof_indices.end());
+        }
+
+    // Get halo layer and accumulate its DoFs
+    std::set<types::global_dof_index> halo_dofs;
+
+    const std::vector<typename DoFHandlerType::active_cell_iterator> halo_cells =
+      GridTools::compute_active_cell_halo_layer(dof_handler, predicate_local);
+    for (typename std::vector<typename DoFHandlerType::active_cell_iterator>::const_iterator it = halo_cells.begin();
+         it != halo_cells.end(); ++it)
+      {
+        // skip halo cells that satisfy the given predicate and thereby
+        // shall be a part of the index set on another MPI core.
+        // Those could only be ghost cells with p::d::Tria.
+        if (predicate(*it))
+          {
+            Assert ((*it)->is_ghost(),
+                    ExcInternalError());
+            continue;
+          }
+
+        const unsigned int dofs_per_cell = (*it)->get_fe().dofs_per_cell;
+        local_dof_indices.resize (dofs_per_cell);
+        (*it)->get_dof_indices (local_dof_indices);
+        halo_dofs.insert(local_dof_indices.begin(),
+                         local_dof_indices.end());
+      }
+
+    // A complication coming from the fact that DoFs living on locally
+    // owned cells which satisfy predicate may participate in constrains for
+    // DoFs outside of this region.
+    if (cm.n_constraints() > 0)
+      {
+        const std::vector<std::pair<types::global_dof_index,double> > *line_ptr;
+
+        // Remove DoFs that are part of constrains for DoFs outside
+        // of predicate. Since we will subtract halo_dofs from predicate_dofs,
+        // achieve this by extending halo_dofs with DoFs to which
+        // halo_dofs are constrained.
+        std::set<types::global_dof_index> extra_halo;
+        for (std::set<types::global_dof_index>::iterator it = halo_dofs.begin();
+             it != halo_dofs.end(); ++it)
+          {
+            line_ptr = cm.get_constraint_entries(*it);
+            // if halo DoF is constrained, add all DoFs to which it's constrained.
+            if (line_ptr!=NULL)
+              {
+                const unsigned int line_size = line_ptr->size();
+                for (unsigned int j=0; j<line_size; ++j)
+                  extra_halo.insert((*line_ptr)[j].first);
```

Comment:
because outside I loop over `halo_dofs` set. I don't want to extend it while looping over it.